# **TikTok Project**
**Course 2 - Get Started with Python**

Welcome to the TikTok Project!

You have just started as a data professional at TikTok.

The team is still in the early stages of the project. You have received notice that TikTok's leadership team has approved the project proposal. To gain clear insights to prepare for a claims classification model, TikTok's provided data must be examined to begin the process of exploratory data analysis (EDA).

A notebook was structured and prepared to help you in this project. Please complete the following questions.

In [1]:
import pandas as pd 
import numpy as np

Then, load the dataset into a dataframe. Creating a dataframe will help you conduct data manipulation, exploratory data analysis (EDA), and statistical activities.

**Note:** As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [2]:
# Load dataset into dataframe
data = pd.read_csv("tiktok_dataset.csv")

### **Task 2b. Understand the data - Inspect the data**

View and inspect summary information about the dataframe by **coding the following:**

1. `data.head(10)`
2. `data.info()`
3. `data.describe()`

*Consider the following questions:*

**Question 1:** When reviewing the first few rows of the dataframe, what do you observe about the data? What does each row represent?

**Question 2:** When reviewing the `data.info()` output, what do you notice about the different variables? Are there any null values? Are all of the variables numeric? Does anything else stand out?

**Question 3:** When reviewing the `data.describe()` output, what do you notice about the distributions of each variable? Are there any questionable values? Does it seem that there are outlier values?

















In [3]:
data.head(10)

# claim_status    video_id  video_duration_sec  \
0   1        claim  7017666017                  59   
1   2        claim  4014381136                  32   
2   3        claim  9859838091                  31   
3   4        claim  1866847991                  25   
4   5        claim  7105231098                  19   
5   6        claim  8972200955                  35   
6   7        claim  4958886992                  16   
7   8        claim  2270982263                  41   
8   9        claim  5235769692                  50   
9  10        claim  4660861094                  45   

                            video_transcription_text verified_status  \
0  someone shared with me that drone deliveries a...    not verified   
1  someone shared with me that there are more mic...    not verified   
2  someone shared with me that american industria...    not verified   
3  someone shared with me that the metro of st. p...    not verified   
4  someone shared with me that the number of busi...    not verified   
5  someone shared with me that gross domestic pro...    not verified   
6  someone shared with me that elvis presley has ...    not verified   
7  someone shared with me that the best selling s...    not verified   
8  someone shared with me that about half of the ...    not verified   
9  someone shared with me that it would take a 50...        verified   

  author_ban_status  video_view_count  video_like_count  video_share_count  \
0      under review          343296.0           19425.0              241.0   
1            active          140877.0           77355.0            19034.0   
2            active          902185.0           97690.0             2858.0   
3            active          437506.0          239954.0            34812.0   
4            active           56167.0           34987.0             4110.0   
5      under review          336647.0          175546.0            62303.0   
6            active          750345.0          486192.0           193911.0   
7            active          547532.0            1072.0               50.0   
8            active           24819.0           10160.0             1050.0   
9            active          931587.0          171051.0            67739.0   

   video_download_count  video_comment_count  
0                   1.0                  0.0  
1                1161.0                684.0  
2                 833.0                329.0  
3                1234.0                584.0  
4                 547.0                152.0  
5                4293.0               1857.0  
6                8616.0               5446.0  
7                  22.0                 11.0  
8                  53.0                 27.0  
9                4104.0               2540.0

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19382 entries, 0 to 19381
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   #                         19382 non-null  int64  
 1   claim_status              19084 non-null  object 
 2   video_id                  19382 non-null  int64  
 3   video_duration_sec        19382 non-null  int64  
 4   video_transcription_text  19084 non-null  object 
 5   verified_status           19382 non-null  object 
 6   author_ban_status         19382 non-null  object 
 7   video_view_count          19084 non-null  float64
 8   video_like_count          19084 non-null  float64
 9   video_share_count         19084 non-null  float64
 10  video_download_count      19084 non-null  float64
 11  video_comment_count       19084 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 1.8+ MB


In [5]:
data.describe()

#      video_id  video_duration_sec  video_view_count  \
count  19382.000000  1.938200e+04        19382.000000      19084.000000   
mean    9691.500000  5.627454e+09           32.421732     254708.558688   
std     5595.245794  2.536440e+09           16.229967     322893.280814   
min        1.000000  1.234959e+09            5.000000         20.000000   
25%     4846.250000  3.430417e+09           18.000000       4942.500000   
50%     9691.500000  5.618664e+09           32.000000       9954.500000   
75%    14536.750000  7.843960e+09           47.000000     504327.000000   
max    19382.000000  9.999873e+09           60.000000     999817.000000   

       video_like_count  video_share_count  video_download_count  \
count      19084.000000       19084.000000          19084.000000   
mean       84304.636030       16735.248323           1049.429627   
std       133420.546814       32036.174350           2004.299894   
min            0.000000           0.000000              0.000000   
25%          810.750000         115.000000              7.000000   
50%         3403.500000         717.000000             46.000000   
75%       125020.000000       18222.000000           1156.250000   
max       657830.000000      256130.000000          14994.000000   

       video_comment_count  
count         19084.000000  
mean            349.312146  
std             799.638865  
min               0.000000  
25%               1.000000  
50%               9.000000  
75%             292.000000  
max            9599.000000

1. Each row contains information about a single video that has been flagged as a claim or opinion. 
2. The data appears to have 19382 rows but some have have null values. Eight out of the twelve columns contain numeric data. The four columns that do not contain numeric data are classified as object.
3. Everything appears to look fairly normal. The longest video in the data is 60 seconds long while the shortest video is 5 seconds. I believe Tiktok allows longer videos, so this data may not be indicative of the entire platform.

### **Task 2c. Understand the data - Investigate the variables**

In this phase, you will begin to investigate the variables more closely to better understand them.

You know from the project proposal that the ultimate objective is to use machine learning to classify videos as either claims or opinions. A good first step towards understanding the data might therefore be examining the `claim_status` variable. Begin by determining how many videos there are for each different claim status.

In [38]:
print('Count of Videos Broken Down by Claim Status:')
data.groupby('claim_status').count()['video_id']

Count of Videos Broken Down by Claim Status:


claim_status
claim      9608
opinion    9476
Name: video_id, dtype: int64

**Question:** What do you notice about the values shown?

There are slightly more claims than opinions, but the dataset is fairly balanced.

Next, examine the engagement trends associated with each different claim status.

Start by using Boolean masking to filter the data according to claim status, then calculate the mean and median view counts for each claim status.

In [27]:
print('Claim Video View Summary Statistics:')
(data.groupby('claim_status').agg(['mean','median'])['video_view_count']).loc['claim']

Claim Video View Summary Statistics


mean      501029.452748
median    501555.000000
Name: claim, dtype: float64

In [28]:
print('Opinion Video View Summary Statistics:')
(data.groupby('claim_status').agg(['mean','median'])['video_view_count']).loc['opinion']

Opinion Video View Summary Statistics:


mean      4956.43225
median    4953.00000
Name: opinion, dtype: float64

**Question:** What do you notice about the mean and median within each claim category?

The mean and median view count is slightly higher for videos containing claims than videos containing opinions.

Now, examine trends associated with the ban status of the author.

Use `groupby()` to calculate how many videos there are for each combination of categories of claim status and author ban status.

In [29]:
print('Video Claim and Author Ban Status Count for Each Category:')
data.groupby(['claim_status', 'author_ban_status']).count()['video_id']

Video Claim and Author Ban Status Count for Each Category:


claim_status  author_ban_status
claim         active               6566
              banned               1439
              under review         1603
opinion       active               8817
              banned                196
              under review          463
Name: video_id, dtype: int64

**Question:** What do you notice about the number of claims videos with banned authors? Why might this relationship occur

There is a considerable number of claim videos belong to banned authors. This leads me to believe that some of the authors made false or miseducated claims that led to their removal from the platform.

Continue investigating engagement levels, now focusing on `author_ban_status`.
Calculate the median video share count of each author ban status.

In [32]:
print('Median Video Share Count for Each Author Ban Status:')
(data.groupby('author_ban_status').agg(['median'])['video_share_count'])

Median Video Share Count for Each Author Ban Status:


median
author_ban_status         
active               437.0
banned             14468.0
under review        9444.0

**Question:** What do you notice about the share count of banned authors, compared to that of active authors? Explore this in more depth.


The median share count for banned authors is much higher than that of active and under review authors. This is probably due to their claim videos that spark controversy leading to more shares and overall engagement.

Use `groupby()` to group the data by `author_ban_status`, then use `agg()` to get the count, mean, and median of each of the following columns:
* `video_view_count`
* `video_like_count`
* `video_share_count`

Remember, the argument for the `agg()` function is a dictionary whose keys are columns. The values for each column are a list of the calculations you want to perform.

In [37]:
data.groupby('author_ban_status').agg(['count', 'mean', 'median'])[['video_view_count', 'video_like_count', 'video_share_count']]

video_view_count                          video_like_count  \
                             count           mean    median            count   
author_ban_status                                                              
active                       15383  215927.039524    8616.0            15383   
banned                        1635  445845.439144  448201.0             1635   
under review                  2066  392204.836399  365245.5             2066   

                                           video_share_count                \
                            mean    median             count          mean   
author_ban_status                                                            
active              71036.533836    2222.0             15383  14111.466164   
banned             153017.236697  105573.0              1635  29998.942508   
under review       128718.050339   71204.5              2066  25774.696999   

                            
                    median  
author_ban_status           
active               437.0  
banned             14468.0  
under review        9444.0

**Question:** What do you notice about the number of views, likes, and shares for banned authors compared to active authors?

Banned authors garner more views, likes, and shares than active and under review author, despite accouting for a small portion of the videos in the larger dataset.

Now, create three new columns to help better understand engagement rates:
* `likes_per_view`: represents the number of likes divided by the number of views for each video
* `comments_per_view`: represents the number of comments divided by the number of views for each video
* `shares_per_view`: represents the number of shares divided by the number of views for each video

In [39]:
data['likes_per_video'] = data['video_like_count']/data['video_view_count']
data['comments_per_video'] = data['video_comment_count']/data['video_view_count']
data['shares_per_view'] = data['video_share_count']/data['video_view_count']

In [41]:
data.head(10)

# claim_status    video_id  video_duration_sec  \
0   1        claim  7017666017                  59   
1   2        claim  4014381136                  32   
2   3        claim  9859838091                  31   
3   4        claim  1866847991                  25   
4   5        claim  7105231098                  19   
5   6        claim  8972200955                  35   
6   7        claim  4958886992                  16   
7   8        claim  2270982263                  41   
8   9        claim  5235769692                  50   
9  10        claim  4660861094                  45   

                            video_transcription_text verified_status  \
0  someone shared with me that drone deliveries a...    not verified   
1  someone shared with me that there are more mic...    not verified   
2  someone shared with me that american industria...    not verified   
3  someone shared with me that the metro of st. p...    not verified   
4  someone shared with me that the number of busi...    not verified   
5  someone shared with me that gross domestic pro...    not verified   
6  someone shared with me that elvis presley has ...    not verified   
7  someone shared with me that the best selling s...    not verified   
8  someone shared with me that about half of the ...    not verified   
9  someone shared with me that it would take a 50...        verified   

  author_ban_status  video_view_count  video_like_count  video_share_count  \
0      under review          343296.0           19425.0              241.0   
1            active          140877.0           77355.0            19034.0   
2            active          902185.0           97690.0             2858.0   
3            active          437506.0          239954.0            34812.0   
4            active           56167.0           34987.0             4110.0   
5      under review          336647.0          175546.0            62303.0   
6            active          750345.0          486192.0           193911.0   
7            active          547532.0            1072.0               50.0   
8            active           24819.0           10160.0             1050.0   
9            active          931587.0          171051.0            67739.0   

   video_download_count  video_comment_count  likes_per_video  \
0                   1.0                  0.0         0.056584   
1                1161.0                684.0         0.549096   
2                 833.0                329.0         0.108282   
3                1234.0                584.0         0.548459   
4                 547.0                152.0         0.622910   
5                4293.0               1857.0         0.521454   
6                8616.0               5446.0         0.647958   
7                  22.0                 11.0         0.001958   
8                  53.0                 27.0         0.409364   
9                4104.0               2540.0         0.183612   

   comments_per_video  shares_per_view  
0            0.000000         0.000702  
1            0.004855         0.135111  
2            0.000365         0.003168  
3            0.001335         0.079569  
4            0.002706         0.073175  
5            0.005516         0.185069  
6            0.007258         0.258429  
7            0.000020         0.000091  
8            0.001088         0.042306  
9            0.002727         0.072714

Use `groupby()` to compile the information in each of the three newly created columns for each combination of categories of claim status and author ban status, then use `agg()` to calculate the count, the mean, and the median of each group.

In [43]:
data.groupby(['claim_status', 'author_ban_status']).agg(['count', 'mean', 'median'])[['likes_per_video', 'comments_per_video', 'shares_per_view']]

likes_per_video                      \
                                         count      mean    median   
claim_status author_ban_status                                       
claim        active                       6566  0.329542  0.326538   
             banned                       1439  0.345071  0.358909   
             under review                 1603  0.327997  0.320867   
opinion      active                       8817  0.219744  0.218330   
             banned                        196  0.206868  0.198483   
             under review                  463  0.226394  0.228051   

                               comments_per_video                      \
                                            count      mean    median   
claim_status author_ban_status                                          
claim        active                          6566  0.001393  0.000776   
             banned                          1439  0.001377  0.000746   
             under review                    1603  0.001367  0.000789   
opinion      active                          8817  0.000517  0.000252   
             banned                           196  0.000434  0.000193   
             under review                     463  0.000536  0.000293   

                               shares_per_view                      
                                         count      mean    median  
claim_status author_ban_status                                      
claim        active                       6566  0.065456  0.049279  
             banned                       1439  0.067893  0.051606  
             under review                 1603  0.065733  0.049967  
opinion      active                       8817  0.043729  0.032405  
             banned                        196  0.040531  0.030728  
             under review                  463  0.044472  0.035027

**Question:**
How does the data for claim videos and opinion videos compare or differ? Consider views, comments, likes, and shares.

Claim videos regardless of ban status get more likes, comments, and shares than opinion videos. When broken down by  author ban status, claim videos perform rougly the same within their cluster and opinion videos perform similarly within their cluster. Overall, claim videos garner more user engagement and this directly correlates to them being flagged.

### **Given your efforts, what can you summarize for Rosie Mae Bradshaw and the TikTok data team?**

*Note for Learners: Your answer should address TikTok's request for a summary that covers the following points:*

*   What percentage of the data is comprised of claims and what percentage is comprised of opinions?
*   What factors correlate with a video's claim status?
*   What factors correlate with a video's engagement level?


## Summary

### Percentage of Claim and Opinion Videos 

In [64]:
claim_mask = (data['claim_status'] == 'claim')
number_of_claims = len(data.loc[claim_mask])
opinion_mask = (data['claim_status'] == 'opinion')
number_of_opinions = len(data.loc[opinion_mask])
print('Percent of Claim Videos: {}%\nPercent of Opinion Videos: {}%'.format(round((number_of_claims*100/len(data)),2),
                                                                                    round((number_of_opinions*100/len(data)),2)
                                                                                   ))

Percent of Claim Videos: 49.57%
Percent of Opinion Videos: 48.89%


### Factors That Correlate to Video Claim Status
* Author Ban Status 
* Video View Count 
* Video Share Count 
* Video Like Count

### Factors That Correlate to Video Engagement
* Author Ban Status 